File: Day-01/go/base2code/day01.go
package main

import (
	"adventofcode-23/utils"
	"fmt"
	"regexp"
	"strconv"
)

var translate = map[string]string{
	"one":   "1",
	"two":   "2",
	"three": "3",
	"four":  "4",
	"five":  "5",
	"six":   "6",
	"seven": "7",
	"eight": "8",
	"nine":  "9",
}

func main() {
	filename := "input/input-1.txt"
	fmt.Println("Part 1:", day1part1(readAllLines(filename)))
}

func day1part1(lines []string) int {
	sum := 0
	for _, line := range lines {
		re := regexp.MustCompile("[0-9]")
		numbers := re.FindAllString(line, -1)
		i1 := utils.StringToInt(numbers[0])
		i2 := utils.StringToInt(numbers[len(numbers)-1])
		str := "" + strconv.Itoa(i1) + strconv.Itoa(i2)
		sum += utils.StringToInt(str)
	}

	return sum
}

func translateIfNecessary(s string) string {
	num, err := strconv.Atoi(s)
	if err != nil {
		return translate[s]
	}
	return strconv.Itoa(num)
}

File: Day-01/ruby/Taucher2003/day_01.rb
input = ARGF.read.lines.map { |line| line.chomp }

def first_and_last_num(str)
  numbers = str.chars.map { |char| Integer(char, exception: false) }.compact
  "#{numbers.first}#{numbers.last}".to_i
end

puts "Part 1: #{input.map(&method(:first_and_last_num)).sum}"

letter_nums = %w[one two three four five six seven eight nine].map
  .with_index(1) { |num, index| [num, "#{num}#{index}#{num}"] }
  .to_h

input_with_numbers = letter_nums.each_with_object(input.dup) do |nums, strs|
  strs.map! { |str| str.gsub(nums.first, nums.last.to_s) }
end

input_with_numbers_map = input_with_numbers.map(&method(:first_and_last_num))

puts "Part 2: #{input_with_numbers_map.sum}"
